packetbeat/protos/drda/drda.go

```diff
@@ -0,0 +1,572 @@
+				parameter.ASCIIData = stripNonASCIIChars(string(data))
+				parameter.EBCDICData = stripNonASCIIChars(string(ebcdic.Decode(data)))
+			}
+
+			m.parameters[codePoint] = *parameter
+			m.RemainingLength -= int(parameterLength)
+			s.parseOffset += int(parameterLength)
+
+			if m.RemainingLength <= 0 {
+				s.parseState = drdaStateDDM
+				return true, true
+			}
+
+			break
+
+		} //end switch
+	} //end for
+
+	return true, false
+}
+
+type drdaPrivateData struct {
+	Data [2]*DrdaStream
+}
+
+func (drda *Drda) ConnectionTimeout() time.Duration {
+	return drda.transactionTimeout
+}
+
+//entry point
+func (drda *Drda) Parse(pkt *protos.Packet, tcptuple *common.TcpTuple,
+	dir uint8, private protos.ProtocolData) protos.ProtocolData {
+
+	trans := drda.getTransaction(tcptuple.Hashable())
+
+	if dir == 1 {
+
+		if trans != nil {
+			logp.Err("transaction should be nil for request")
+		}
+
+		trans = &DrdaTransaction{Type: "drda", tuple: *tcptuple, TsStart: pkt.Ts}
+		drda.transactions.Put(tcptuple.Hashable(), trans)
+		logp.Debug("drdadetailed", "Initialize transaction")
+
+	} else {
+		if trans == nil {
+			logp.Err("transaction should be not nil for response")
+		}
+	}
+
+	//dir == 1 request
+	//dir == 0 response
+
+	//relevant tcp packet
+
+	defer logp.Recover("ParseDrda exception")
+
+	priv := drdaPrivateData{}
+	if private != nil {
+		var ok bool
+		priv, ok = private.(drdaPrivateData)
+		if !ok {
+			priv = drdaPrivateData{}
+		}
+	}
+
+	if priv.Data[dir] == nil {
+		priv.Data[dir] = &DrdaStream{
+			tcptuple: tcptuple,
+			data:     pkt.Payload,
+			message:  &DrdaMessage{},
+		}
+	} else {
+		// concatenate bytes
+		priv.Data[dir].data = append(priv.Data[dir].data, pkt.Payload...)
+		if len(priv.Data[dir].data) > tcp.TCP_MAX_DATA_IN_STREAM {
+			logp.Debug("drda", "Stream data too large, dropping TCP stream")
+			priv.Data[dir] = nil
+			return priv
+		}
+	}
+
+	completed := true
+
+	stream := priv.Data[dir]
+	for len(stream.data) > 0 {
+		if stream.message == nil {
+			stream.message = &DrdaMessage{}
+		}
+
+		stream.message.Direction = dir
+
+		ok, complete := drdaMessageParser(priv.Data[dir])
+		//logp.Debug("drdadetailed", "drdaMessageParser returned ok=%b complete=%b", ok, complete)
+		if !ok {
+			// drop this tcp stream. Will retry parsing with the next
+			// segment in it
+			priv.Data[dir] = nil
+			logp.Debug("drdadetailed", "Ignore DRDA message. Drop tcp stream. Try parsing with the next segment")
+			return priv
+		}
+
+		if complete {
+
+			stream.message.TcpTuple = *tcptuple
+			stream.message.Direction = dir
+			stream.message.CmdlineTuple = procs.ProcWatcher.FindProcessesTuple(tcptuple.IpPort())
+
+			if stream.message.Direction == 1 {
+				drda.receivedDrdaRequest(stream.message)
+			} else {
+				drda.receivedDrdaResponse(stream.message)
+			}
+
+			// and reset message
+			stream.PrepareForNewMessage()
+		} else {
+			// wait for more data
+			completed = false
+			break
+		}
+	}
+
+	if completed {
```

Comment:
for loop only returns if buffer empty or incomplete message buffered. What happens if 2 message are buffered? Why not have the body of this if inside the loop?